File: golabs/gocloud/godata/go-gumball-riak/gumball-riak/src/github.com/basho/taste-of-riak/nodejs/Ch03-Msgy-Schema/timeline-manager.js
'use strict';

var Msg = require('./models/msg');
var Timeline = require('./models/timeline');

function TimelineManager(argTimelineRepository, argMsgRepository) {
    this.timelineRepository = argTimelineRepository;
    this.msgRepository = argMsgRepository;

    this.addToTimeline = function (msg, timelineType, msgKey, callback) {
        var self = this,
            timelineKey = this.generateKeyFromMsg(msg, timelineType);

        self.timelineRepository.get(timelineKey, true, function (err, timeline) {
            if (timeline) {
                timeline = addToExistingTimeline(timeline, msgKey);
            } else {
                timeline = createNewTimeline(timelineKey, msgKey);
            }

            self.timelineRepository.save(timeline, callback);
        });
    }

    this.generateKey = function (ownerUsername, timelineType, date) {
        if (ownerUsername && timelineType && date) {
            var dateString = date.toISOString().substring(0, 10);
            return ownerUsername + "_" + timelineType + "_" + dateString;
        } else {
            throw new Error(
                "ownerUsername: " + ownerUsername + ", " +
                "timelineType: " + timelineType + ", " +
                "date: " + date);
        }
    }

    this.generateKeyFromMsg = function (msg, timelineType) {
        var owner = getOwner(msg, timelineType);
        return this.generateKey(owner, timelineType, msg.created);
    }

    function addToExistingTimeline(timeline, msgKey) {
        timeline.addMsg(msgKey);
        return timeline;
    }

    function createNewTimeline(timelineKey, msgKey) {
        var newTimeline = new Timeline(timelineKey);
        newTimeline.addMsg(msgKey);
        return newTimeline;
    }

    function getOwner(msg, timelineType) {
        switch (timelineType)
        {
            case Timeline.TimelineType.Inbox:
                return msg.recipient;
            case Timeline.TimelineType.Sent:
                return msg.sender;
        }
        return msg.recipient;
    }
}

TimelineManager.prototype.postMsg = function (msg, callback) {
    var self = this;
    self.msgRepository.save(msg, function (err, savedMsg) {
        if (err) {
            callback(err, null);
        }
        // Post to recipient's Inbox timeline
        self.addToTimeline(msg, Timeline.TimelineType.Inbox, savedMsg.id, function (err, recipientRslt) {
            if (err) {
                callback(err, null);
            }
            // Post to sender's Sent timeline
            self.addToTimeline(msg, Timeline.TimelineType.Sent, savedMsg.id, function (err, senderRslt) {
                if (err) {
                    callback(err, null);
                } else {
                    callback(null, [ recipientRslt, senderRslt ]);
                }
            });
        });
    });
};

TimelineManager.prototype.getTimeline = function (ownerUsername, timelineType, date, callback) {
    var timelineKey = this.generateKey(ownerUsername, timelineType, date);
    this.timelineRepository.get(timelineKey, false, callback);
}

module.exports = TimelineManager;

File: golabs/gocloud/godata/go-gumball-riak/gumball-riak/src/github.com/basho/taste-of-riak/csharp/Ch03-Msgy-Schema/TimelineManager.cs
namespace Ch03_Msgy_Schema
{
    using System;
    using Models;
    using Repositories;
    using RiakClient;

    public class TimelineManager
    {
        private readonly IRepository<Timeline> timelineRepository;
        private readonly IRepository<Msg> msgRepository;

        public TimelineManager(IRepository<Timeline> timelineRepository, IRepository<Msg> msgRepository)
        {
            if (timelineRepository == null)
            {
                throw new ArgumentNullException("timelineRepository");
            }
            this.timelineRepository = timelineRepository;

            if (msgRepository == null)
            {
                throw new ArgumentNullException("msgRepository");
            }
            this.msgRepository = msgRepository;
        }

        public void PostMsg(Msg msg)
        {
            string msgKey = msgRepository.Save(msg);

            // Post to recipient's Inbox timeline
            AddToTimeline(msg, Timeline.TimelineType.Inbox, msgKey);

            // Post to sender's Sent timeline
            AddToTimeline(msg, Timeline.TimelineType.Sent, msgKey);
        }

        public Timeline GetTimeline(string ownerUsername, Timeline.TimelineType type, DateTime date)
        {
            string timelineKey = GenerateKey(ownerUsername, type, date);
            return timelineRepository.Get(timelineKey);
        }

        private void AddToTimeline(Msg msg, Timeline.TimelineType type, string msgKey)
        {
            string timelineKey = GenerateKeyFromMsg(msg, type);

            Timeline timeline = timelineRepository.Get(key: timelineKey, notFoundOK: true);
            if (timeline != null)
            {
                timeline = AddToExistingTimeline(timeline, msgKey);
            }
            else
            {
                timeline = CreateNewTimeline(timelineKey, msgKey);
            }

            timelineRepository.Save(timeline);
        }

        private static Timeline AddToExistingTimeline(Timeline timeline, string msgKey)
        {
            timeline.AddMsg(msgKey);
            return timeline;
        }

        private static Timeline CreateNewTimeline(string timelineKey, string msgKey)
        {
            var newTimeline = new Timeline(timelineKey);
            newTimeline.AddMsg(msgKey);
            return newTimeline;
        }

        private string GenerateKeyFromMsg(Msg msg, Timeline.TimelineType type)
        {
            string owner = GetOwner(msg, type);
            return GenerateKey(owner, type, msg.Created);
        }

        private static string GetOwner(Msg msg, Timeline.TimelineType type)
        {
            switch (type)
            {
                case Timeline.TimelineType.Inbox:
                    return msg.Recipient;
                case Timeline.TimelineType.Sent:
                    return msg.Sender;
            }

            return msg.Recipient;
        }

        private static string GenerateKey(string ownerUsername, Timeline.TimelineType type, DateTime date)
        {
            string dateString = date.ToString("yyyy-MM-dd");
            return ownerUsername + "_" + type.ToString() + "_" + dateString;
        }
    }
}
